ap/production/gen_top_decay.py

```diff
@@ -0,0 +1,153 @@
+    #
+    # groups = np.concatenate((
+    #     t[[[0, 1], [0], [0, 1, 2], ...]][:, :, None],
+    #     w[[[1, 0], [0], [0, 2, 1], ...]][:, :, None],
+    #     ...
+    # ), axis=2)
+    #
+    # the object slices, e.g. [[1, 0], [0], [0, 2, 1], ...] for w, account for the mapping between
+    # the objects and for that matter, t dictate the order and therefore actually have no slicing
+    # the other particles, w, b, qs and ls, however, use them to match "their" top and
+    # the [:, :, None] just adds a new axis that is required for concatenation
+    # no grouping algorithm implemented yet for 3 or more t/W/b or for 2 with same charge
+    # note: in case we can verify that gen particles are stored depth-first, then any matching of
+    #       b/w/qs/ls to t would be trivial as they would be intrinsically in the same order
+    sign = lambda part: (part.pdgId > 0) * 2 - 1
+    t_sign = sign(t)
+    all_or_raise((nt != 2) | (ak.sum(t_sign, axis=1) == 0), "grouping not implemented for 2 ss tops, but found")
+    all_or_raise(nt <= 2, "grouping not implemented for 3 or more tops, but found")
+    mask2 = nt == 2
+
+    # for w, start with the local index as is and for events with 2 objects, order using sign of the
+    # pdg id so that it matches that of the top quark
+    w_idxs = ak.local_index(w, axis=1)
+    w_sign = sign(w)
+    w_flip = mask2 & (ak.sum(t_sign * w_sign, axis=1) < 0)
+    if ak.any(w_flip):
+        w_idxs = ak.where(w_flip, w_idxs[:, ::-1], w_idxs)
+
+    # for b, do the same as for w
+    b_idxs = ak.local_index(b, axis=1)
+    b_flip = mask2 & (ak.sum(t_sign * sign(b), axis=1) < 0)
+    if ak.any(b_flip):
+        b_idxs = ak.where(b_flip, b_idxs[:, ::-1], b_idxs)
+
+    # for quarks and leptons, first build pairs under the assumption that two consecutive
+    # objects make up a W (checked later)
+    qs_pairs = ak.concatenate([qs[:, ::2, None], qs[:, 1::2, None]], axis=2)
+    ls_pairs = ak.concatenate([ls[:, ::2, None], ls[:, 1::2, None]], axis=2)
+
+    # reorder within pairs so that d-type quark, or charged lepton is at the front (odd pdg number)
+    def idxs_odd_pdg_id_first(pairs):
+        pairs_idxs = ak.local_index(pairs, axis=2)
+        pairs_flip = abs(pairs.pdgId)[:, :, 0] % 2 == 0
+        if ak.any(pairs_flip):
+            pairs_idxs = ak.where(pairs_flip, pairs_idxs[:, :, ::-1], pairs_idxs)
+        return pairs_idxs
+
+    qs_pairs = qs_pairs[idxs_odd_pdg_id_first(qs_pairs)]
+    ls_pairs = ls_pairs[idxs_odd_pdg_id_first(ls_pairs)]
+
+    # merge pairs and then order to match the correct W per pair, again using a sign comparison
+    w_prod = ak.concatenate([qs_pairs, ls_pairs], axis=1)
+    w_prod_sign = ak.flatten(sign(w_prod[(w_prod.pdgId % 2 == 0)]), axis=2)
+    w_prod_flip = (w_sign != w_prod_sign)[:, 0]
+    if ak.any(w_prod_flip):
+        w_prod = ak.where(w_prod_flip, w_prod[:, ::-1], w_prod)
+
+    # create the groups
+    groups = ak.concatenate(
```

Comment:
Just because it's quite hard to digest, the output would look like:

`[[top1, W1, b1, [q1, q2]], [top2, ...], ...]`

for a tt event, correct?